man/man1/gen_nodes_file.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH GEN_NODES_FILE 1 "October 12, 2012"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
gen_nodes_file - Generate a nodes file for input to
.B pbs_cluster_status
.SH SYNOPSIS
.B gen_nodes_file
.RI [ options ]
.br
.SH DESCRIPTION
.B gen_nodes_file
generates an initial "nodes" file for use with
.B pbs_cluster_status.
.PP
HPC cluster systems are often split into several sub-clusters each
controlled by the same batch system.  Usually there are so many nodes, that
each node in a given sub-cluster has a specific prefix
(e.g. 'mynode-n001', 'mynode-n002', ... etc.).
.B gen_nodes_file
assumes this structure and expects as input an XML file generated from the
output of the 'pbsnodes -x' command.  One then specifies a section name for
each of the sub-clusters, a node prefix for the nodes in the given sub-cluster
and a table width (e.g. to split a 96-node cluster into 6 rows of 16 one would
use a table width of 16).
.B gen_nodes_file
uses this information to generate a 'nodes' file in the required format
for visualisation with the
.B pbs_cluster_status
program of the
.B pbsclusterviz
project.
.SH OPTIONS
This program follows the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
.TP
.B \-h, \-\-help
Print usage information and exit
.TP
.B \-v, \-\-version
Print version information and exit
.TP
.B \-x, \-\-xmlfile=<filename>
Specify an input pbsnodes xml file
.TP
.B \-o, \-\-nodesfile=<filename>
Specify an output nodes file
.TP
.B \-n, \-\-section_name=<name>
Specify a name for the node section
.TP
.B \-p, \-\-node_prefix=<prefix>
Specify a nodename prefix
.TP
.B \-w, \-\-table_width=<width>
Number of nodes across a section
.TP
.B \-a, \-\-append
Append to existing nodes file?
.TP
.br
The program is documented fully by the 
.B pbsclusterviz
manual, which should be available at https://github.com/paultcochrane/pbsclusterviz.
.SH AUTHOR
.B gen_nodes_file
was written by Paul Cochrane <paul@liekut.de>.
.SH COPYRIGHT
Copyright \(co 2009 Free Software Foundation, Inc.
License GPLv2+: GNU GPL version 2 or later
<http://gnu.org/licenses/gpl.html>.
.br
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
.SH SEE ALSO
pbs_cluster_status(1), http://pbsclusterviz.sf.net
.PP
This manual page was written by Paul Cochrane <paul@liekut.de>,
for the Debian project (but may be used by others).


CHANGES
0.7a
====

 - merge of "interactive" branch with trunk
 - pbsnodes file can be downloaded via ssh and viewed locally on the user's
   computer
 - added a command line option to specify the ssh command if desired
 - log information about overloaded/down/offline nodes now displayed
 - display automatically updated
 - added configuration options for pbsnodes file download, display update,
   and what to display in log information
 - pop-up balloons appear with node information after pressing the 'i' key
   in interactive mode
 - pop-up balloons require python-vtk 5.2+
 - nodes not appearing in nodes file but which appear in pbsnodes output
   (and vice versa) are detected and mentioned in error output
 - documentation update; better integration of docs with sourceforge website
 - examples updated to agree with current code
 - several minor bug fixes

0.6a
====

 - Improvement of internal structure
 - gen_nodes_file works again
 - Removed old, unused classes and files
 - Improved and extended debugging capabilities
 - Default display mode is now 'job'
 - gen_nodes_file now says what it is doing instead of being silent
 - Setting the display mode on the command line now works
 - Documentation updated
 - Colourmap changes properly between 'load' and 'job' view
 - Display updates with the 'u' key now works
 - The -c/--configfile command line option performs now as expected
 - The nodes_file command line option now works

0.5a
====

 - Prerelease of the new interactive visualisation capabilities.
 - Interactive output is now default.
 - cluster_load_status and cluster_job_status consolidated into one file
 - Keyboard input directly interpreted to change between "load" and "job"
   display modes.
 - Balloons pop up (with the "i" key) to show the jobs running on a given
   node and the total load on that node.
 - Many internal structural changes.

0.1
===

 - First public release.


gen_nodes_file.py
#!/usr/bin/env python

# Copyright (C) 2009 Paul Cochrane
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
# USA.

"""
Automatically generate a nodes file
"""

import pbsclusterviz
from pbsclusterviz.pbs.pbsnodes import PBSNodes, Node
from pbsclusterviz.pbs.xml_handler import PBSNodesXMLHandler
from pbsclusterviz.compute_node import ComputeNode
import xml.sax
import getopt
import sys, re


def usage():
    print """Usage:
    python gen_nodes_file.py [options]

    options:
    [-h/--help]                     Print usage information and exit
    [-V/--version]                  Print version information and exit
    [-x/--xmlfile=<filename>]       Specify an input xml file
    [-o/--nodesfile=<filename>]     Specify an nodes file
    [-n/--section_name=<name>]      Specify a prefix for the node section
    [-p/--section_prefix=<prefix>]  Specify a prefix for the node section
    [-w/--table_width=<width>]      Number of nodes across a section
    [-a/--append]                   Append to existing nodes file?
    """

def version():
    print """gen_nodes_file version 0.1a"""

#---------------------------------------------------------------------
# Handle options
#---------------------------------------------------------------------

try:
    options_list, args_list = getopt.getopt(sys.argv[1:], "hVax:o:n:p:w:d",
            ["help", "version", "xmlfile=", "nodesfile=", 
		"section_name=", "section_prefix=", "table_width=",
		"append", "debug"])
except getopt.GetoptError:
    # print help information and exit:
    usage()
    sys.exit(2)

xml_file = None
nodes_file = "nodes.gen"
pbsclusterviz.pbs.__debug = False
section_name = None
section_prefix = None
table_width = 10
append_to_nodes_file = False
for option, arg in options_list:
    if option in ("-h", "--help"):
        usage()
        sys.exit()
    if option in ("-V", "--version"):
        version()
        sys.exit()
    elif option in ("-x", "--xmlfile"):
        xml_file = arg
    elif option in ("-o", "--outfile"):
        nodes_file = arg
    elif option in ("-n", "--section_name"):
	section_name = arg
    elif option in ("-p", "--section_prefix"):
	section_prefix = arg
    elif option in ("-w", "--table_width"):
	table_width = int(arg)
    elif option in ("-a", "--append"):
        pbsclusterviz.pbs.__debug = True
    elif option in ("-d", "--debug"):
        pbsclusterviz.pbs.__debug = True
    else:
        print "Unknown option %s" % option
        sys.exit(2)

if len(args_list) > 2:
    print "Too many arguments"
    usage()
    sys.exit()

if section_name is None:
    print "You need to specify a section name (-n <name>)"
    sys.exit(2)

if section_prefix is None:
    print "You need to specify a section prefix (-p <prefix>)"
    sys.exit(2)

pbsnodes = PBSNodes()
parser = xml.sax.make_parser()
handler = PBSNodesXMLHandler(pbsnodes)
parser.setContentHandler(handler)

if xml_file is not None:
    # TODO: check that the file exists
    parser.parse(xml_file)
else:
    # TODO: implement a direct call to pbsnodes -x
    print "Direct call to pbsnodes not yet implemented"
    sys.exit()

# the section prefix is used to filter the different kinds of nodes on a
# multi-cluster system
prefix_regex = re.compile(r"^%s" % section_prefix)

# TODO: build in the ability to automatically determine the current y grid
# position from a previous nodes file.  This should only be done if the
# "append" option is used.

x_grid_pos = 0
y_grid_pos = 0

fp = None
if append_to_nodes_file:
    fp = open(nodes_file, "a")
else:
    fp = open(nodes_file, "w")

fp.write("# section: %s\n" % section_name)
for node in pbsnodes.get_node_list():
    node_name = node.get_name()
    if prefix_regex.match(node_name):
        fp.write("%s %i %i\n" % (node_name, x_grid_pos, y_grid_pos))
	# if we're at the side of the table, reset x and increment y
	if x_grid_pos >= table_width-1:
	    x_grid_pos = 0
	    y_grid_pos += 1
	else:
	    x_grid_pos += 1

fp.close()


